logtail/logtail_test.go

```diff
@@ -23,28 +27,213 @@ func TestFastShutdown(t *testing.T) {
+			if err != nil {
+				t.Error("failed to read HTTP request")
+			}
+			ts.uploaded <- body
 		}))
-	defer testServ.Close()
 
-	l := NewLogger(Config{BaseURL: testServ.URL}, t.Logf)
-	for i := 1; i < 10; i++ {
+	t.Cleanup(ts.srv.Close)
+	return &ts
+}
+
+func (ts *LogtailTestServer) HandleStartLoggingMessage() {
+	// There is always an initial "logtail started" message
+	body := <-ts.uploaded
+	if !strings.Contains(string(body), "started") {
+		panic("unknown start logging statement:" + string(body))
+	}
+}
+
+func TestDrainPendingMessages(t *testing.T) {
+	ts := NewLogtailTestServer(t)
+	l := NewLogger(Config{BaseURL: ts.srv.URL}, t.Logf)
+	ts.HandleStartLoggingMessage()
+
+	for i := 0; i < logLines; i++ {
 		l.Write([]byte("log line"))
 	}
 
-	l.Shutdown(ctx)
-	cancel()
-	if uploads == 0 {
-		t.Error("no log uploads")
+	// all of the "log line" messages usually arrive at once, but poll if needed.
+	body := ""
+	for i := 0; i <= logLines; i++ {
+		body += string(<-ts.uploaded)
+		cnt := strings.Count(body, "log line")
+		if cnt == logLines {
+			break
+		}
+	}
+
+	err := l.Shutdown(context.Background())
+	if err != nil {
+		t.Error(err)
+	}
+
+	if strings.Count(body, "log line") != logLines {
+		t.Errorf("got %q; want: %d copies of %q", body, logLines, "log line")
+	}
+}
+
+func TestEncodeAndUploadMessages(t *testing.T) {
+	ts := NewLogtailTestServer(t)
+	l := NewLogger(Config{BaseURL: ts.srv.URL}, t.Logf)
+	ts.HandleStartLoggingMessage()
+
+	tests := []struct {
+		name string
+		log  string
+		want string
+	}{
+		{
+			"plain text",
+			"log line",
+			"log line",
+		},
+		{
+			"simple JSON",
+			`{"text": "log line"}`,
+			"log line",
+		},
+	}
+
+	for _, tt := range tests {
+		io.WriteString(l, tt.log)
+		body := <-ts.uploaded
+
+		data := make(map[string]interface{})
+		err := json.Unmarshal(body, &data)
+		if err != nil {
+			t.Error(err)
+		}
+
+		got, ok := data["text"]
+		if ok {
+			if got != tt.want {
+				t.Errorf("%s: got %q; want %q", tt.name, got.(string), tt.want)
+			}
+		} else {
+			t.Errorf("no text in: %v", data)
+		}
+
+		ltail, ok := data["logtail"]
+		if ok {
+			logtailmap := ltail.(map[string]interface{})
+			_, ok = logtailmap["client_time"]
+			if !ok {
+				t.Errorf("%s: no client_time present", tt.name)
+			}
+		} else {
+			t.Errorf("%s: no logtail map present", tt.name)
+		}
+	}
+
+	err := l.Shutdown(context.Background())
+	if err != nil {
+		t.Error(err)
+	}
+}
+
+func TestEncodeSpecialCases(t *testing.T) {
+	ts := NewLogtailTestServer(t)
+	l := NewLogger(Config{BaseURL: ts.srv.URL}, t.Logf)
+	ts.HandleStartLoggingMessage()
+
+	// -------------------------------------------------------------------------
+
+	// JSON log message already contains a logtail field.
+	io.WriteString(l, `{"logtail": "LOGTAIL", "text": "text"}`)
+	body := <-ts.uploaded
+	data := make(map[string]interface{})
+	err := json.Unmarshal(body, &data)
+	if err != nil {
+		t.Error(err)
+	}
+	error_has_logtail, ok := data["error_has_logtail"]
```

Comment:
Go uses camelCase instead of underscore_case.

Reply:
Done